proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,257 @@
+
+As a result:
+
+* Reduced contribution and development velocity, prolonging bugs, important features and efficiency efforts.
+* Increased operational cost and reduced trust for users who run highly durable Prometheus setups.
+
+### Context: TSDB format changes
+
+TSDB format, including chunk format (also used in the head with mmapped chunks), is also a handwritten variadic-encoding binary format. Compared to WAL it's generally versioned e.g.
+
+* [Chunk disk file format](https://github.com/prometheus/prometheus/blob/12c39d5421cc29a4bfc13fc57fd9ccd3dbc310f0/tsdb/docs/format/chunks.md#L14).
+* [Head chunk disk format](https://github.com/prometheus/prometheus/blob/fd62dbc2918deea0ceae94758baf7a095b52dd5b/tsdb/docs/format/head_chunks.md#L13).
+* [Index file](https://github.com/prometheus/prometheus/blob/d699dc3c7706944aafa56682ede765398f925ef0/tsdb/docs/format/index.md#L8).
+* [Block meta file](https://github.com/prometheus/prometheus/blob/5e124cf4f2b9467e4ae1c679840005e727efd599/tsdb/block.go#L171).
+* Chunks have its own ["encoding type"](https://github.com/prometheus/prometheus/blob/a693dd19f244f000da40bcbac85041846b78cfc1/tsdb/chunkenc/chunk.go#L29) that could be used for versioning.
+
+No docs or formal strategy was developed, but Prometheus generally follow the ["2-fold" migration strategy explained below](#two-fold-migration-strategy).
+
+## Goals
+
+* Agree on the official recommendations for **lossless WAL migrations** strategies for devs and users.
+* Balancing development velocity with user data stability risks.
+
+## Non-Goals
+
+* TSDB format can but does not need to follow the same recommendations. We want to change the WAL format now, so prioritizing WAL.
+* Mentioning Write-Before-Log, it uses WAL format so all apply to WBL too.
+* To reduce scope we don't mention [memory snapshot format](https://github.com/prometheus/prometheus/blob/fd5ea4e0b590d57df1e2ef41d027f2a4f640a46c/tsdb/docs/format/memory_snapshot.md#L1) for now.
+
+## How
+
+We recommend the [Two-Fold Migration Strategy](#two-fold-migration-strategy) with two details:
+
+* A new flag that tells Prometheus what WAL version to write.
+* There can be multiple "forward compatible" version, but the official minimum is one (see, the rejected [LTS idea](#require-lts-support))
+
+We propose to add a string `--storage.tsdb.stateful.write-wal-version` flag, with the default to `v1` that has a "stateful" consequence -- once new version is used, users will be able to revert only to certain Prometheus versions. Help of this flag will explain clearly what's possible and what Prometheus version you will be able to revert to.
+ 
+In other words, we propose to add a TSDB flag `--storage.tsdb.stateful.write-wal-version=(v1 or v2)` that tells Prometheus to use a particular WAL format for both WAL and WBL. This kind of flag will change its default to a new version ONLY when (at least) one previous Prometheus version can read that version (while writing the old one).
```

Comment:
I've never seen such notation, looks weird, let's use what you get when you run `prometheus --help`.
```suggestion
In other words, we propose to add a TSDB flag `--storage.tsdb.stateful.write-wal-version=<version>` that tells Prometheus to use a particular WAL format for both WAL and WBL. This kind of flag will change its default to a new version ONLY when (at least) one previous Prometheus version can read that version (while writing the old one). The initial version would be `v1`.
```

Reply:
That's what I meant (: thx.